neural_compressor/adaptor/torch_utils/teq.py

```diff
@@ -0,0 +1,391 @@
+#
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2023 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+from .smooth_quant import GraphTrace, get_module, set_module
+from .weight_only import quant_weight
+from .model_wrapper import TEQLinearFakeQuant, TEQMulLinear
+import torch
+import logging
+from transformers import get_scheduler
+
+logging.basicConfig(
+        format="%(asctime)s - %(levelname)s - %(name)s - %(message)s",
+        datefmt="%m/%d/%Y %H:%M:%S",
+        level=logging.INFO)
+
+logger = logging.getLogger(__name__)
+
+
+class TEQuantizer:
+    """
+    Weight-only quantization, Trainable Equivalent Transformation (TEQ): linear wrapper to apply scale to input
+    """
+
+    def __init__(
+        self,
+        model,
+        weight_config={},
+        example_inputs=None
+    ):
+        """
+        :param model: the model for quantization
+        :param weight_config (dict, optional): contains all info required by GPTQ. Defaults to {}.
+        :param example_inputs: inputs for trace 
+        """
+
+        self.model = model
+        self.num_bits = weight_config.get('wbits', 4)
+        self.group_size = weight_config.get('group_size', -1)
+        self.scheme = weight_config.get('sym', False)
+        self.folding = weight_config.get('folding', True)
+        self.example_inputs = example_inputs
+        self.device, self.dtype = self._get_device()
+        self.model.eval()
+
+        self.trained_alphas = {}
+
+    def _get_device(self):
+        """
+        Get the model device
+        :return:Model device
+        """
+        for _, p in self.model.named_parameters():
+            return p.data.device, p.data.dtype
+
+    def add_tuning_scale(self, op_types=['Linear', 'Conv2d'], excluded_name="lm_head",
+            excluded_key=None, sqrt_w_init=False):
+        """
+        The main entry of smooth quant
+        to the paper for more details
+        :param op_types: The op typed to be smooth quantized
+        :param excluded_name: exclude layer
+        :param excluded_key: exclude key
+        :param sqrt_w_init: use sqrt weight to init
```

Comment:
need to support the transformer.conv1d later

Reply:
ok